File: 0003_length_of_longest_substring.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        maximum = 0
        start = 0
        end = 0
        dictionary = {}
        for i in range(len(s)):
            if s[i] in dictionary and dictionary[s[i]] + 1 > start:
                start = dictionary[s[i]] + 1
            dictionary[s[i]] = end
            end += 1
            maximum = max(maximum, end - start)
        return maximum

   
# Alternative solution

class Solution1(object):
    
    def lengthOfLongestSubstring(self, s):
        maximum = 0
        string = ''
        for i in range(len(s)):
            print(string)
            if s[i] not in string:
                string += s[i]
                maximum = max(maximum, len(string))
            else:
                string = string[string.index(s[i]) + 1:]
                string += s[i]
        return maximum

File: 0001-0500/0003_lengthOfLongestSubstring.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashSet<Character> window = new HashSet<>();
        int start = 0;
        int end = 0;
        int longest = 0;

        while (end < s.length()) {
            char c = s.charAt(end);

            while (window.size() > 0 && window.contains(c)) {
                window.remove(s.charAt(start));
                start++;
            }

            window.add(c);
            longest = Math.max(longest, window.size());
            end++;
        }

        return longest;
    }
}
